File: Stage25/Python/Problem12015.py
import sys
input = sys.stdin.readline

def main():
    N = int(input())
    arr = list(map(int, input().split()))

    lis = []

    for x in arr:
        left, right = 0, len(lis)

        while left < right:
            mid = (left + right) // 2
            if lis[mid] < x:
                left = mid + 1
            else:
                right = mid

        if left == len(lis):
            lis.append(x)
        else:
            lis[left] = x

    print(len(lis))

if __name__ == "__main__":
    main()

File: Stage25/C/Problem12015.c
#include <stdio.h>

int lower_bound(int arr[], int size, int target) {
    int left = 0, right = size;

    while (left < right) {
        int mid = (left + right) / 2;
        if (arr[mid] < target)
            left = mid + 1;
        else
            right = mid;
    }
    return left;
}

int main(void) {
    int N;
    scanf("%d", &N);

    int lis[N];
    int size = 0;

    for (int i = 0; i < N; i++) {
        int x;
        scanf("%d", &x);

        int idx = lower_bound(lis, size, x);
        lis[idx] = x;

        if (idx == size)
            size++;
    }

    printf("%d\n", size);
    return 0;
}